docs/user/tutorial-asgi.rst

```diff
@@ -0,0 +1,752 @@
+      "modified": "Tue, 24 Dec 2019 17:32:19 GMT",
+      "size": [
+          462,
+          462
+      ]
+  }
+
+Accessing the newly uploaded image::
+
+  $ http localhost:8000/images/5cfd9fb6-259a-4c72-b8b0-5f4c35edcd3c.jpeg
+
+  HTTP/1.1 200 OK
+  content-type: image/jpeg
+  date: Tue, 24 Dec 2019 17:34:53 GMT
+  server: uvicorn
+  transfer-encoding: chunked
+
+  +-----------------------------------------+
+  | NOTE: binary data not shown in terminal |
+  +-----------------------------------------+
+
+We could also open the link in the web browser to verify the converted JPEG
+image looks as intended.
+
+Let's check the image collection now::
+
+  $ http localhost:8000/images
+
+  HTTP/1.1 200 OK
+  content-length: 175
+  content-type: application/json
+  date: Tue, 24 Dec 2019 17:36:31 GMT
+  server: uvicorn
+
+  [
+      {
+          "id": "5cfd9fb6-259a-4c72-b8b0-5f4c35edcd3c",
+          "image": "/images/5cfd9fb6-259a-4c72-b8b0-5f4c35edcd3c.jpeg",
+          "modified": "Tue, 24 Dec 2019 17:32:19 GMT",
+          "size": [
+              462,
+              462
+          ]
+      }
+  ]
+
+The application file layout should now look like::
+
+  asgilook
+  ├── .venv
+  └── asgilook
+      ├── __init__.py
+      ├── app.py
+      ├── asgi.py
+      ├── config.py
+      ├── images.py
+      └── store.py
+
+Dynamic Thumbnails
+------------------
+
+Let's pretend our image service customers want to render images in multiple
+resolutions, for instance, as ``srcset`` for responsive HTML images or other
+purposes.
+
+Let's add a new method ``Store.make_thumbnail()`` to perform scaling on the
+fly:
+
+.. code:: python
+
+    async def make_thumbnail(self, image, size):
+        async with aiofiles.open(image.path, 'rb') as img_file:
+            data = await img_file.read()
+
+        loop = asyncio.get_running_loop()
+        return await loop.run_in_executor(None, self._resize, data, size)
+
+As well as an internal helper to run the ``Pillow`` thumbnail operation that
+is offloaded to a threadpool executor, again, in hoping that Pillow can release
+the GIL for some operations:
+
+.. code:: python
+
+    def _resize(self, data, size):
+        image = PIL.Image.open(io.BytesIO(data))
+        image.thumbnail(size)
+
+        resized = io.BytesIO()
+        image.save(resized, 'JPEG')
+        return resized.getvalue()
+
+The ``store.Image`` class can be extended to also return URIs to thumbnails:
+
+.. code:: python
+
+    def thumbnails(self):
+        def reductions(size, min_size):
+            width, height = size
+            factor = 2
+            while width // factor >= min_size and height // factor >= min_size:
+                yield (width // factor, height // factor)
+                factor *= 2
+
+        return [
+            f'/thumbnails/{self.image_id}/{width}x{height}.jpeg'
+            for width, height in reductions(
+                self.size, self.config.min_thumb_size)]
```

Comment:
The reader has not yet been asked to add `min_thum_size` to their config.'

`AttributeError: 'Config' object has no attribute 'min_thumb_size'`

Reply:
Ouch :man_facepalming: I really need to retest the whole flow once again after all that refactoring.